### PR TITLE
Framework for document processors to restructure TOC

### PR DESCRIPTION
**Pull request by vicancy**
So that each Processor can register delegates to restructure TOC in its own processor

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/HostService.cs [hunk 1]
```diff
@@ -123,6 +124,43 @@ public MarkupResult Markup(string markdown, FileAndType ft, bool omitParse)
...
+        {
+            if (methods == null || methods.Length == 0)
+            {
+                return;
+            }
+
+            var list = new List<RestructureTableOfContent>(methods);
+            lock (_tocSyncRoot)
+            {
+                if (RestructureTableOfContentDelegates != null)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/HostService.cs]: >RestructureTableOfContentDelegates [](start = 20, length = 34)

it won't be null? has initialized?

**Comment by qinezh** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/HostService.cs (hunk 1 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/HostService.cs]: It is public accessible, so it can be set to null by callers.

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/HostService.cs]: get it, so it's better also check null in ShouldRestructureTableOfContent and InvokeRestructuringTableOfContent?

---
In reply to: [93891813](https://github.com/dotnet/docfx/pull/1059#discussion_r93891813) [](ancestors = 93891813)

File: src/Microsoft.DocAsCode.Build.Engine/CompilePhaseHandler.cs [hunk 2]
```diff
@@ -38,8 +40,25 @@ public void Handle(List<HostService> hostServices, int maxParallelism)
                     {
                         Prebuild(hostService);
                     }
+
+                    // Register all the delegates to handler
+                    if (hostService.RestructureTableOfContentDelegates != null)
+                    {
+                        _delegates.AddRange(hostService.RestructureTableOfContentDelegates);
+                    }
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/CompilePhaseHandler.cs]: how about put these into Prepare()?

**Comment by ansyral** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/CompilePhaseHandler.cs [hunk 3]
```diff
@@ -38,8 +40,25 @@ public void Handle(List<HostService> hostServices, int maxParallelism)
                     {
                         Prebuild(hostService);
                     }
+
+                    // Register all the delegates to handler
+                    if (hostService.RestructureTableOfContentDelegates != null)
+                    {
+                        _delegates.AddRange(hostService.RestructureTableOfContentDelegates);
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/CompilePhaseHandler.cs]: >AddRange [](start = 35, length = 8)

question: if same delegate are added twice, we should invode twice?

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/CompilePhaseHandler.cs (hunk 3 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/CompilePhaseHandler.cs]: yes, hostservice decides

**Comment by hellosnow** [comment]
:shipit:

**Pull request closed without merging**